### PR TITLE
Mapping experiments

### PR DESCRIPTION
**Pull request by Nurkambay**

### CODE REVIEW

**Review by ishche — APPROVED** [review]
LGTM

File: server/engine/src/main/java/org/eclipse/lsp/cobol/core/engine/MappingService.java [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+ */
+package org.eclipse.lsp.cobol.core.engine;
+
+import lombok.Getter;
+import lombok.Value;
+import org.eclipse.lsp4j.Location;
+import org.eclipse.lsp4j.Position;
+import org.eclipse.lsp4j.Range;
+
+import java.util.*;
```
Comment (lxpdd) [inline server/engine/src/main/java/org/eclipse/lsp/cobol/core/engine/MappingService.java:23]: there aren't too many imports from java.util, could we mention only those we use?

File: server/engine/src/main/java/org/eclipse/lsp/cobol/core/engine/dialects/idms/IdmsCopybookDescriptor.java [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+ *    : literal | cobolWord
+ *    ;
+ */
+
+/**
+ * Description of IDMS copybook
+ */
+@Data
+public class IdmsCopybookDescriptor {
+  String name;
```
Comment (lxpdd) [inline server/engine/src/main/java/org/eclipse/lsp/cobol/core/engine/dialects/idms/IdmsCopybookDescriptor.java:52]: please don't break encapsulation rules here, make those fields private

File: server/engine/src/main/java/org/eclipse/lsp/cobol/core/engine/dialects/idms/IdmsDialect.java [hunk 3]
```diff
@@ -41,45 +52,154 @@ public final class IdmsDialect implements CobolDialect {
...
+    CopyNode copyNode = new CopyNode(cb.getStatement(), cb.getName(), IdmsDialect.NAME);
+    if (recursiveCall(copybookStack, copyNode.getName())) {
+      textTransformations.replace(copyNode.getLocality().getRange(), "");
+      errors.add(ErrorHelper.circularDependency(messageService, cb.getUsage(), cb.getName()));
+      return;
+    }
+    copybookStack.push(copyNode.getName());
+
+    Location cbLocation = new Location();
+    cbLocation.setRange(new Range(new Position(0, 0), new Position(0, 0)));
```
Comment (lxpdd) [inline server/engine/src/main/java/org/eclipse/lsp/cobol/core/engine/dialects/idms/IdmsDialect.java:108]: Position(0,0) could be replaced with Position(),
was found 10 times in this PR

File: server/engine/src/main/java/org/eclipse/lsp/cobol/core/engine/TextTransformations.java [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+    extensions.put(range, textTransformations);
+  }
+
+  /**
+   * Substitute location of original document with a new content.
+   *
+   * @param range location
+   * @param newText new content
+   */
+  public void replace(Range range, String newText) {
```
Comment (lxpdd) [inline server/engine/src/main/java/org/eclipse/lsp/cobol/core/engine/TextTransformations.java:159]: that method is too often being called with newText = "",
so do consider to add a new method to remove all the text from that range.

File: server/engine/src/main/java/org/eclipse/lsp/cobol/core/engine/dialects/idms/IdmsCopybookDescriptor.java (hunk 2 above)
Comment (ishche) [reply server/engine/src/main/java/org/eclipse/lsp/cobol/core/engine/dialects/idms/IdmsCopybookDescriptor.java:52]: They are, it's Lombok magic

Comment (lxpdd) [reply server/engine/src/main/java/org/eclipse/lsp/cobol/core/engine/dialects/idms/IdmsCopybookDescriptor.java:52]: I doubt that, the thing that lombok does with `@Data` is creating getters and setters (and something more), but it's possible to threat that as dumb structure with `descriptor.name =`

File: server/engine/src/main/java/org/eclipse/lsp/cobol/core/engine/dialects/idms/IdmsCopybookDescriptor.java [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+   * @return copybook descriptor
+   */
+  public static IdmsCopybookDescriptor from(IdmsCopyParser.CopyIdmsStatementContext ctx, String programDocumentUri) {
+    IdmsCopybookDescriptor result = new IdmsCopybookDescriptor();
+    result.name = getName(ctx);
+    Range range = DialectUtils.constructRange(ctx.copyIdmsOptions().copyIdmsSource());
+
+    if (ctx.LEVEL_NUMBER() != null && ctx.LEVEL_NUMBER().getSymbol() != null) {
+      result.levelRange = extractLevelRange(ctx.LEVEL_NUMBER().getSymbol());
+      result.level = extractLevel(ctx.LEVEL_NUMBER().getSymbol());
```
Comment (lxpdd) [inline server/engine/src/main/java/org/eclipse/lsp/cobol/core/engine/dialects/idms/IdmsCopybookDescriptor.java:71]: @ishche `result.level =` lombok magic ? 🧙

**Comment by VitGottwald** [comment]
Close?

**Comment by lxpdd** [comment]
> Close?

Yes, it could be closed

**Pull request closed without merging**